metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationDriver.java

```diff
@@ -391,13 +391,16 @@ void enqueueMetadataChangeEvent(
 
     // Events handled by Migration Driver.
     abstract class MigrationEvent implements EventQueue.Event {
+        // Use no-op handler by default because the retryHandler will be overridden if needed
+        public void retryHandler() { }
         @SuppressWarnings("ThrowableNotThrown")
         @Override
         public void handleException(Throwable e) {
             if (e instanceof MigrationClientAuthException) {
                 KRaftMigrationDriver.this.faultHandler.handleFault("Encountered ZooKeeper authentication in " + this, e);
             } else if (e instanceof MigrationClientException) {
                 log.info(String.format("Encountered ZooKeeper error during event %s. Will retry.", this), e.getCause());
+                retryHandler();
```

Comment:
I feel the `retry` is existent except for `UNINITIALIZED` since `UNINITIALIZED` is not running by another event. For other event type, `PollEvent` will put (do-something event + one deferred `PollEvent`) to the queue. It means the deferred `PollEvent` is the "retry".

My question is "why we did not handle `UNINITIALIZED` by another event"? If we move `recoverMigrationStateFromZK` to another event, we don't need to add extra `retryHandler`.

Reply:
Also, the solution offered by this PR has a side effect that we will put 2 `PollEvent` if the exception `MigrationClientException` happens in other `migrationState`

Reply:
> My question is "why we did not handle UNINITIALIZED by another event"? If we move recoverMigrationStateFromZK to another event, we don't need to add extra retryHandler.

That's a good quesiton, @chia7712 ! Let me think about it.

> Also, the solution offered by this PR has a side effect that we will put 2 PollEvent if the exception MigrationClientException happens in other migrationState

No, as you said above, the `MigrationClientException` won't happen in other migrationState because they will be handled in other event handler, which is not related to pollEvent. And because the default retryHandler is no-op, there will be no retry for other migrationStates. As long as `pollEvent` is keep polling, they can be retried later.

Reply:
@chia7712 , I take your suggestion to add `RecoverMigrationStateFromZKEvent` so that we don't need to worry about retry anymore. I was checking if this change will cause any side effect, and here is my finding:
1. `recoverMigrationStateFromZK` is expected to run before the driver starts the state machine.
2. In the `recoverMigrationStateFromZK`, we'll do these things:
  a. create a ZNode for migration and initial migration state
  b. install this class as a metadata publisher
  c. transition to INACTIVE state
3. If this `recoverMigrationStateFromZK` is keep failing, the log will keep outputting errors and keep retrying. Once it succeeds, the metadata publisher will be installed and the `onControllerChange` and `onMetadataUpdate` will be triggered to start the process. That means, if we change `recoverMigrationStateFromZK` into an event, it won't affect anything because what we need to do at this state is just waiting for the (a)(b)(c) operation completes.

So, I'm +1 with this suggestion. Thank you.